src/v/kafka/server/handlers/fetch.cc

```diff
@@ -66,6 +70,17 @@ make_partition_response_error(model::partition_id p_id, error_code error) {
     };
 }
 
+namespace {
+model::timeout_clock::time_point
+failsafe_deadline_from(const model::timeout_clock::time_point& t) {
+    if (safeguard_fetch_timeout > model::no_timeout - t) {
+        return t;
```

Comment:
nit: if the intention of failsafe_dealine_from is to perform a saturating_add, this should return model::no_timeout
otherwise we might have some weird behavior if some code downstream checks for equality against model::no_timeout (but it's a remote possibility)

Reply:
Thanks for pointing out `saturating_add`, did not know about this earlier but yes, looking at it's definition, that is what we are trying to do here. 

I don't understand the following statement though, say we are not able to safely add the safeguard and just return `t`.  The most common case will be when `t == no_timeout` and there we end up returning `no_timeout`. 

But in other cases, where `t` is within 5 minutes of `no_timeout`, we return `t`. What issues might this cause? There is no invariant or assumption that the `failsafe_deadline` is supposed to be equal to `no_timeout` in some code paths.

Returning `no_timeout` here also is fine, just want to understand what the advantages of that would be.

Reply:
you are right that it's not an issue, if this path is taken it means that it was huge anyway. 
it's just a matter that I recognize returning no_timeout to mean "this is an addition with some protection for overflows", while returning t makes me pause a second.
I'm also thinking if in the future we want to specialize the reader in case of timeout==model::no_timeout, maybe for logging purposes or stuff like that.
but it's a nit anyway

Reply:
I changed it to return `model::no_timeout` as I think the corner case of `t` being within 5 minutes of `model::no_timeout` is going to be rare enough that we do not want to bother with special casing, if returning `model::no_timeout` is easier to understand then it is a good reason to change this.